File: dotnet/DesignPatterns/CreationalDesignPatterns/AbstractFactoryDesignPattern/AbstractFactoryDesignPattern.Example1/Application.cs
namespace AbstractFactoryDesignPattern.Example1;

public class Application
{
	private readonly GUIFactory _factory;

	private Button _button;
	private Menu _menu;
	private Dialog _dialog;
	public Application(GUIFactory factory)
	{
		_factory = factory;
	}

	public void CreateGUI()
	{
		_button = _factory.CreateButton();
		_menu = _factory.CreateMenu();
		_dialog = _factory.CreateDialog();
	}

	public void RunApplication()
	{
		Console.WriteLine("=== Starting GUI Application ===");

		_button.Render();
		_menu.Render();
		_dialog.Render();

		Console.WriteLine("\n=== User Interaction ===");

		_menu.AddItem("File");
		_menu.AddItem("Edit");
		_menu.AddItem("Appearance");

		_dialog.Show();
		_button.OnClick();
		_dialog.Close();
	}
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/AbstractFactoryDesignPattern/AbstractFactoryDesignPattern.Example1/src/Application.ts
import { GUIFactory } from './factory/abstract/GUIFactory';
import { Button } from './models/abstract/Button';
import { Menu } from './models/abstract/Menu';
import { Dialog } from './models/abstract/Dialog';

export class Application {
  private button!: Button;
  private menu!: Menu;
  private dialog!: Dialog;

  constructor(private readonly factory: GUIFactory) {}

  createGUI(): void {
    this.button = this.factory.createButton();
    this.menu = this.factory.createMenu();
    this.dialog = this.factory.createDialog();
  }

  runApplication(): void {
    console.log('=== Starting GUI Application ===');

    this.button.render();
    this.menu.render();
    this.dialog.render();

    console.log('\n=== User Interaction ===');

    this.menu.addItem('File');
    this.menu.addItem('Edit');
    this.menu.addItem('Appearance');

    this.dialog.show();
    this.button.onClick();
    this.dialog.close();
  }
}
